azure/services/privateendpoints/spec.go

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package privateendpoints
+
+import (
+	"context"
+
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2022-05-01/network"
+	"github.com/pkg/errors"
+	"k8s.io/utils/pointer"
+	infrav1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	"sigs.k8s.io/cluster-api-provider-azure/azure/converters"
+)
+
+// PrivateEndpointSpec defines the specification for a private endpoint.
+type PrivateEndpointSpec struct {
+	Name                       string
+	ResourceGroup              string
+	Location                   string
+	CustomNetworkInterfaceName string
+	PrivateIPAddress           string
+	SubnetID                   string
+	PrivateLinkServiceID       string
+	GroupIds                   []string
+	ApplicationSecurityGroup   string
+	RequestMessage             string
+	ManualApproval             bool
+	AdditionalTags             infrav1.Tags
+	ClusterName                string
+}
+
+// ResourceName returns the name of the private endpoint.
+func (s *PrivateEndpointSpec) ResourceName() string {
+	return s.Name
+}
+
+// ResourceGroupName returns the name of the resource group.
+func (s *PrivateEndpointSpec) ResourceGroupName() string {
+	return s.ResourceGroup
+}
+
+// OwnerResourceName is a no-op for private endpoints.
+func (s *PrivateEndpointSpec) OwnerResourceName() string {
+	return ""
+}
+
+// Parameters returns the parameters for the PrivateEndpointSpec.
+func (s *PrivateEndpointSpec) Parameters(ctx context.Context, existing interface{}) (interface{}, error) {
+	if existing != nil {
+		if _, ok := existing.(network.PrivateEndpoint); !ok {
+			return nil, errors.Errorf("%T is not a network.PrivateEndpoint", existing)
+		}
+		// PrivateEndpointSpec already exists
+		return nil, nil
```

Comment:
If we don't intend for the controller to make changes to these resources after they're first created, I think it would be good to add some validation in the webhooks that enforces these values in the CAPZ resources can't be changed. Otherwise I think any changes would silently no-op which users probably wouldn't expect.

Reply:
Refactored it so that updates are taken into consideration too. Not sure yet which fields are immutable so I prefer letting Azure API fails rather than adding any sort of restrictions at the webhook level. Over time we might reconsider it once we'll get more inputs over which fields are immutable in private endpoint spec.